datafusion/core/src/physical_plan/aggregates/row_hash.rs

```diff
@@ -474,7 +646,97 @@ impl GroupedHashAggregateStream {
+                .iter_mut()
+                .any(|x| x.try_reserve(&batch).is_err())
+                || self.group_values.try_reserve(&batch).is_err()
+                || self
+                    .current_group_indices
+                    .try_reserve(batch.num_rows())
+                    .is_err()
+                || self.update_memory_reservation().is_err()
+                || self.force_spill;
+            if should_spill {
+                // If mode is Final, use input batch (Partial mode) schema for spilling because
+                // the spilled data will be merged and re-evaluated later.
+                if let AggregateMode::Final | AggregateMode::FinalPartitioned = self.mode
+                {
+                    self.spill_schema = batch.schema();
+                }
+                self.spill()?;
+                self.clear_shrink(batch);
+
+                return self.update_memory_reservation();
+            }
+        }
+        Ok(())
+    }
+
+    /// Emit all rows, sort them, and store them on disk.
+    fn spill(&mut self) -> Result<()> {
+        let emit = self.emit(EmitTo::All, true)?;
+        let sorted = sort_batch(&emit, &self.spill_expr, None)?;
+        let spillfile = self.runtime.disk_manager.create_tmp_file("HashAggSpill")?;
+        let mut writer = IPCWriter::new(spillfile.path(), &emit.schema())?;
+        // TODO: slice large `sorted` and write to multiple files in parallel
+        writer.write(&sorted)?;
+        writer.finish()?;
+        self.spills.push(spillfile);
+        Ok(())
+    }
+
+    /// Clear memory and shirk capacities to the size of the batch.
+    fn clear_shrink(&mut self, batch: &RecordBatch) {
+        self.accumulators
+            .iter_mut()
+            .for_each(|x| x.clear_shrink(&batch));
+        self.group_values.clear_shrink(&batch);
+        self.current_group_indices.clear();
+        self.current_group_indices.shrink_to(batch.num_rows());
+    }
+
+    /// At this point, all the inputs are read and there are some spills.
+    /// Emit the remaining rows and create a batch.
+    /// Conduct a streaming merge sort between the batch and spilled data. Since the stream is fully
+    /// sorted, set `self.group_ordering` to Full, then later we can read with [`EmitTo::First`].
+    fn update_merged_stream(&mut self) -> Result<()> {
+        let batch = self.emit(EmitTo::All, true)?;
+        let mut streams: Vec<SendableRecordBatchStream> = vec![];
+        let expr = self.spill_expr.clone();
+        let schema = batch.schema();
+        streams.push(Box::pin(RecordBatchStreamAdapter::new(
+            schema.clone(),
+            futures::stream::once(futures::future::lazy(move |_| {
+                sort_batch(&batch, &expr, None)
+            })),
+        )));
+        for spill in self.spills.drain(..) {
+            let stream = read_spill_as_stream(spill, schema.clone())?;
+            streams.push(stream);
+        }
+        self.merged_stream = streaming_merge(
```

Comment:
Instead of merge-sorting all spills first and then aggregating, which is a two-pass procedure, I suggest we directly merge the aggregation states by keys and produce output at the same time. 

A similar example could be found in `physical_plan/aggregates/order/full.rs`, the only difference is the number of input streams while aggregating.

Reply:
Here the merge sort is streaming, so it does not immediately merge all spills. In that way, we can take batch by batch from the top of the stream whenever the data is ready. Then we can reduce the memory pressure.
Yes I am using `GroupOrderingFull` here (couple lines below.)

Reply:
Thanks, that makes sense.